include/rcutils/time.h

```diff
@@ -36,7 +36,10 @@ extern "C"
 #define RCUTILS_US_TO_NS(microseconds) ((microseconds) * 1000LL)
 
 /// Convenience macro to convert nanoseconds to seconds.
-#define RCUTILS_NS_TO_S(nanoseconds) ((nanoseconds) / (1000LL * 1000LL * 1000LL))
+#define RCUTILS_NS_TO_S(nanoseconds) \
+  ((1e-9 * (double)((int32_t)(nanoseconds % 1000000000l))) + \
+  ((double)((int32_t)((nanoseconds - ((int32_t)(nanoseconds % 1000000000l))) / 1000000000l))))
```

Comment:
See my comment in https://github.com/ros2/rmw_fastrtps/pull/755 on why I'm very nervous about making this change.  Basically things that were expecting integer division before are now going to get FPU division.  While it may be more accurate, it also changes the expectations.

This actually leads to another observation, which is that the type of `nanoseconds` is ill-defined because this is a macro.

All of that leads me to another thought.  Maybe what we should do here is to make a new function called `double rcutils_nanoseconds_to_seconds(double)`.  And then we (slowly) convert uses of the macro over to that new function.  That way we aren't making this change for things that don't expect it, and we can get the types right on the new function.

Thoughts?

Reply:
> While it may be more accurate, it also changes the expectations.

Yeah as we found in [this review](https://github.com/ros2/rcutils/pull/460/files#r1573147454), changing the nature of the macro induces errors in testing. To be fair, it's a beneficial change, but if it causes errors in testing there's a high chance it could lead to breakages on the users' end. Though I don't think going the route of new typed function is a better option as it just adds more to the file and alters the way code is written, when I would assume users just want to use a macro, design-wise. Unless most of the macros are rebranded to typed functions, I would feel that it's just an awkward change when it comes to DX.

> This actually leads to another observation, which is that the type of nanoseconds is ill-defined because this is a macro.

Yeah while [rcutils](https://github.com/ros2/rcutils/blob/cf3e9986a557e4c893eb718b90c77ce189494a11/include/rcutils/time.h#L48) and [dds implementations](https://github.com/eProsima/Fast-DDS/blob/8cb447bb25032f646685c923da866b733fd78aff/src/cpp/rtps/common/Time_t.cpp#L114) use `int64_t` it's not necessarily common knowledge for users.

I would like to create more accuracy, but since there's not a effective way to deprecate macros, I doubt it's worth breaking user behavior to do so, nor is it worth expanding on.

Reply:
> I would like to create more accuracy, but since there's not a effective way to deprecate macros, I doubt it's worth breaking user behavior to do so, nor is it worth expanding on.

I think that adding more accurate function calls (especially if they are constexpr) and migrating where it makes sense would work?

Reply:
> I think that adding more accurate function calls (especially if they are constexpr) and migrating where it makes sense would work?

Yeah, that's what I'm thinking we should do.

Reply:
> (especially if they are constexpr) and migrating where it makes sense would work?

Seems fair enough to make more calls, but since `rcutils` is essentially C repository, did we need these split functions in a different repo?

Reply:
> Seems fair enough to make more calls, but since `rcutils` is essentially C repository, did we need these split functions in a different repo?

Good point.  It can't be constexpr here, but it can at least be a function, be const, and use types.

Alternatively, if we wanted to use C++ and constexpr, we could move it to `rcpputils`, but then there are certain places it can't be used (like `rcutils`, `rcl`, `rcl_action`, and `rmw`).